feat(docs): Add a split keyboards feature page (#2406)
zmkfirmware/zmk: docs/docs/features/split-keyboards.md

However, if this process somehow went wrong or you used controllers in a different split keyboard configuration before, you will need to explicitly clear the stored bond information so that the parts can pair properly.
For this, please follow [the specified procedure](../troubleshooting/connection-issues.mdx#split-keyboard-halves-unable-to-pair) in the troubleshooting section.

## Behaviors with Locality

ZMK [behaviors](../behaviors/index.mdx) are typically processed and affect the central part of the split keyboard since it handles the keymap state and communication with the host devices.
However, certain behaviors have different "localities" and need to necessarily affect the peripheral: these include behaviors with global and source localities.

### Global Locality Behaviors

These are behaviors that affect features on all keyboard parts and where the state needs to be explicitly synced between them. Currently these are the following behaviors:

- [RGB underglow behaviors](../behaviors/underglow.md)
- [Backlight behaviors](../behaviors/backlight.md)
- [Power management behaviors](../behaviors/power.md)

> - [Soft off behavior](../behaviors/soft-off.md)

Author: This should be updated if #2330 gets merged first.

---
title: Split Keyboards
sidebar_label: Split Keyboards
---

> ZMK supports split keyboard setups where two or more keyboard parts (also called "sides" or "halves"), each with their own controller running ZMK, communicate with each other to work as a single keyboard device.

Author: I had actually used "part" as the generic term, then added the sides/halves remark at the last minute because they are used throughout the rest of the docs. Let me know if you have a better term in mind than "part," I'll think about it as well.

> ZMK supports split keyboard setups where two or more keyboard parts (also called "sides" or "halves"), each with their own controller running ZMK, communicate with each other to work as a single keyboard device.

Reviewer: Perhaps "unit" or "node" might work nicely? We could also stick with part, but I'm worried about that being too generic.

> ZMK supports split keyboard setups where two or more keyboard parts (also called "sides" or "halves"), each with their own controller running ZMK, communicate with each other to work as a single keyboard device.

Author: Both sound a bit too technical to me, so I think I prefer "parts" over them. I'd be interested to hear if anyone else has preferences.

> ZMK supports split keyboard setups where two or more keyboard parts (also called "sides" or "halves"), each with their own controller running ZMK, communicate with each other to work as a single keyboard device.

Reviewer: "pieces", perhaps?

:::note[Split communication protocols]
Currently ZMK only supports split keyboards that communicate with each other wirelessly over BLE.
As such, only controllers that support BLE can be used with ZMK split keyboards.

Supporting split communication over wired protocols is planned, which would allow using ZMK split keyboards with non-wireless controllers.
:::

## Central and Peripheral Roles

In ZMK splits, one keyboard part has the "central" role that receives key position and sensor events from the other parts that are called "peripherals."
This part runs the necessary keymap logic to convert received events into HID events such as keycodes and then communicates with the connected host devices, e.g. over USB or bluetooth.

> As a result of this architecture the internal keyboard state (like active layers) is shared and is mainly handled by the central part.

Author: I think I was thinking about underglow/ext power etc. state since they just synced from the central. Even then, I think it would be fair to say "exclusively handled by the central" and would simplify understanding.

> As a result of this architecture the internal keyboard state (like active layers) is shared and is mainly handled by the central part.

Author: I think I want to preserve the intention of "is shared" here, where I was trying to emphasize the "work as a single keyboard device" from above. Any suggestions on wording that better?

> As a result of this architecture the internal keyboard state (like active layers) is shared and is mainly handled by the central part.

Author: On a second look I like the suggested version.

However this also means that the peripheral parts _cannot_ communicate with host devices on their own, since they can only communicate with the central part.
They will not present as keyboard devices when connected over USB and will not advertise as pairable BLE keyboards.

:::note[Latency considerations]
Since peripherals communicate through centrals, the key and sensor events originating from them will naturally have a larger latency, especially with a wireless split communication protocol.

> :::

Author: I think that makes sense. In that case I should convert this to a subsection perhaps?

> :::

Reviewer: I think that would be nice, yeah.

> :::

Author: I added a section but I omitted any references to settings because
- these are ZMK internal Kconfig symbols that currently aren't documented in the config section
- they are already set for minimum possible latency so I didn't see much point in giving pointers to change them

For instance when [using the GitHub workflow](../user-setup.mdx) to build two part split keyboards, two firmware files that typically contain `<keyboard>_left` and `<keyboard>_right` in the file names will be produced.
These files need to be flashed to the respective controllers of the two halves.

Since the keymap processing is primarily done on the central side, for keymap changes it will typically be enough to flash the controller of the central half.
However if you make changes to [config files](../config/index.md) that should apply to all parts, you need to flash to all parts.
Any changes in ZMK related to split keyboard features might also necessitate doing this.

## Pairing for Wireless Split Keyboards

Split keyboards with BLE-based split communications (which are all split keyboards that are officially supported) have an internal pairing procedure between the central and each peripheral.
When the central has an open slot for a peripheral, it will advertise for connections (which will not be visible to non-ZMK devices).
Then, any peripheral that has not yet bonded to a peripheral will pair to it.

> Similar to how [bluetooth profiles](bluetooth.md) are managed between the keyboard and host devices, the bonding information will be stored with the corresponding hardware addresses of the other keyboard part, on both the central and peripheral.

Author: Good one, I'll add a warning box here.

However, if this process somehow went wrong or you used controllers in a different split keyboard configuration before, you will need to explicitly clear the stored bond information so that the parts can pair properly.
For this, please follow [the specified procedure](../troubleshooting/connection-issues.mdx#split-keyboard-halves-unable-to-pair) in the troubleshooting section.

## Behaviors with Locality

ZMK [behaviors](../behaviors/index.mdx) are typically processed and affect the central part of the split keyboard since it handles the keymap state and communication with the host devices.
However, certain behaviors have different "localities" and need to necessarily affect the peripheral: these include behaviors with global and source localities.

### Global Locality Behaviors

These are behaviors that affect features on all keyboard parts and where their state needs to be explicitly synced between them. Currently these are the following behaviors:

- [RGB underglow behaviors](../behaviors/underglow.md)
- [Backlight behaviors](../behaviors/backlight.md)
- [Power management behaviors](../behaviors/power.md)

> - [Soft off behavior](../behaviors/soft-off.md)

Author: Actually the "central" locality is the default, but maybe that wasn't very clear from the preceding paragraph? (Most behaviors are central, where they don't communicate anything to peripherals.)

> - [Soft off behavior](../behaviors/soft-off.md)

Reviewer: I see, that wasn't very clear no. Taking that into account, I think [some of the content here](https://github.com/zmkfirmware/zmk/pull/2406/files#r1703661628) should be pulled into this section and then everything can be made more explicit.

> - [Soft off behavior](../behaviors/soft-off.md)

Author: I tried to rewrite this a bit to make it more intuitive and less technical. I don't know if it is any clearer for you but if you have a suggestion let me know.

### Source Locality Behaviors

These behaviors only affect the keyboard part that they are invoked from, given that they were invoked from a plain behavior binding (i.e. not nested inside another behavior) on a keymap layer. These behaviors include:

- [Reset behaviors](../behaviors/reset.md)

:::note[Peripheral invocation]
The peripheral side of the keyboard has to be paired and connected to the central side in order to be able to activate these behaviors, even if it is possible to trigger the behavior using only keys on that side. This is because the key bindings are processed on the central side which would then instruct the peripheral side to reset.

> :::

Author: Not sure myself, but we can keep it in mind for experimenting in the future.

## Behaviors with Locality

ZMK [behaviors](../behaviors/index.mdx) are typically processed and affect the central part of the split keyboard since it handles the keymap state and communication with the host devices.
However, certain behaviors have different "localities" and need to necessarily affect the peripheral: these include behaviors with global and source localities.

### Global Locality Behaviors

These are behaviors that affect features on all keyboard parts and where their state needs to be explicitly synced between them. Currently these are the following behaviors:

- [RGB underglow behaviors](../behaviors/underglow.md)
- [Backlight behaviors](../behaviors/backlight.md)
- [Power management behaviors](../behaviors/power.md)
- [Soft off behavior](../behaviors/soft-off.md)

### Source Locality Behaviors

These behaviors only affect the keyboard part that they are invoked from, given that they were invoked from a plain behavior binding (i.e. not nested inside another behavior) on a keymap layer. These behaviors include:

- [Reset behaviors](../behaviors/reset.md)

:::note[Peripheral invocation]

> The peripheral side of the keyboard has to be paired and connected to the central side in order to be able to activate these behaviors, even if it is possible to trigger the behavior using only keys on that side. This is because the key bindings are processed on the central side which would then instruct the peripheral side to reset.

Author: ```suggestion
The peripheral side of the keyboard has to be paired and connected to the central side in order to be able to activate these behaviors, even if it is possible to trigger the behavior using only keys on that side. This is because the key bindings are processed on the central side which would then instruct the peripheral side to run the behavior's effect.
```
Copy paste mishap.

Peripherals _cannot_ communicate with host devices on their own, since they can only communicate with the central part.
They will not present as keyboard devices when connected over USB and will not advertise as pairable BLE keyboards.

By convention, for two part split keyboards left "halves" are set as centrals and right halves as peripherals.

### Configuration

The [new shield guide](../development/new-shield.mdx) details how to define a split keyboard shield with two parts, enabling the split feature and setting up the necessary roles for each part.

Also see the reference section on [split keyboards configuration](../config/system.md#split-keyboards) where the relevant symbols include `CONFIG_ZMK_SPLIT` that enables the feature, `CONFIG_ZMK_SPLIT_ROLE_CENTRAL` which sets the central role and `CONFIG_ZMK_SPLIT_BLE_CENTRAL_PERIPHERALS` that sets the number of peripherals.

### Latency Considerations

Since peripherals communicate through centrals, the key and sensor events originating from them will naturally have a larger latency, especially with a wireless split communication protocol.

> For the currently used BLE-based transport, the average latency added by the split communication is 3.75ms with a worst case latency of 7.5ms.

Reviewer: ```suggestion
For the currently used BLE-based transport, split communication increases the average latency by 3.75ms with a worst case increase of 7.5ms.
```

Similar to how [bluetooth profiles](bluetooth.md) are managed between the keyboard and host devices, the bonding information will be stored with the corresponding hardware addresses of the other keyboard part, on both the central and peripheral.

In practice, this means that your split keyboard parts will automatically pair and work the first time they are all on at the same time.
However, if this process somehow went wrong or you used controllers in a different split keyboard configuration before, you will need to explicitly clear the stored bond information so that the parts can pair properly.
For this, please follow [the specified procedure](../troubleshooting/connection-issues.mdx#split-keyboard-halves-unable-to-pair) in the troubleshooting section.

:::warning
If the central keyboard part is either advertising for a pairing or waiting for disconnected peripherals, it will consume more power and drain batteries faster.
:::

## Behaviors with Locality

Most ZMK [behaviors](../behaviors/index.mdx) are processed exclusively on the central part of the split keyboard since it handles the keymap state and any communication with the host devices.

> However, certain behaviors have "global" or "source" localities, where they can have an effect the peripheral parts when invoked.

Reviewer: ```suggestion
However, certain behaviors have "global" or "source" localities, where they can affect the peripheral parts when invoked.
```

### Global Locality Behaviors

> These are behaviors that have an effect on all keyboard parts, such as changing lighting effects.

Reviewer: ```suggestion
These are behaviors that affect on all keyboard parts, such as changing lighting effects.
```

Currently these are the following behaviors:

- [RGB underglow behaviors](../behaviors/underglow.md)
- [Backlight behaviors](../behaviors/backlight.md)
- [Power management behaviors](../behaviors/power.md)
- [Soft off behavior](../behaviors/soft-off.md)

### Source Locality Behaviors

> These behaviors only affect the keyboard part that they are invoked from, given that they were invoked from a plain behavior binding (i.e. not nested inside another behavior) on a keymap layer. These behaviors include:

Reviewer: What happens if you invoke reset from a tap dance?

Author: This is referring to #1494 and yeah, it would apply to tap dances as well. On second thought, since this causes issues with the global behaviors as well I might move this to a warning box that applies to both?

Reviewer: I think that would be good. Maybe get rid of the "Given the were invoked" part, and move all of that into a warning box which talks about nesting (and perhaps also list the nested behaviours we have? If that ends up looking nice)

Author: I didn't put a full list but hopefully covered the common cases. Uncommon cases would be sensor rotate, perhaps macro (I haven't tested this one)

---
title: Split Keyboards
sidebar_label: Split Keyboards
---

> ZMK supports split keyboard setups where two or more keyboard parts (also called "sides" or "halves"), each with their own controller running ZMK, communicate with each other to work as a single keyboard device.

Reviewer: ```suggestion
ZMK supports setups where a keyboard is split into two or more physical parts (also called "sides" or "halves" when split in two), each with their own controller running ZMK. The parts communicate with each other to work as a single keyboard device.
```

:::note[Split communication protocols]
Currently ZMK only supports split keyboards that communicate with each other wirelessly over BLE.
As such, only controllers that support BLE can be used with ZMK split keyboards.

Supporting split communication over wired protocols is planned, allowing for ZMK split keyboards using non-wireless controllers.
:::

## Central and Peripheral Roles

> In ZMK splits, one keyboard part has the "central" role that receives key position and sensor events from the other parts that are called "peripherals."

Reviewer: ```suggestion
In split keyboards running ZMK, one part is assigned the "central" role which receives key position and sensor events from the other parts that are called "peripherals."
```

> This part runs the necessary keymap logic to convert received events into HID events such as keycodes and then communicates with the connected host devices, e.g. over USB or bluetooth.

Reviewer: ```suggestion
The central runs the necessary keymap logic to convert received events into HID events such as keycodes and then communicates with the connected host devices, e.g. over USB or bluetooth.
```

> The internal keyboard state (like active layers) is handled exclusively by the central part.

Reviewer: ```suggestion
The internal keyboard state (like active layers) is handled exclusively by the central.
```

> Peripherals _cannot_ communicate with host devices on their own, since they can only communicate with the central part.

Reviewer: ```suggestion
Peripherals _cannot_ communicate with host devices on their own, since they can only communicate with the central.
```

They will not present as keyboard devices when connected over USB and will not advertise as pairable BLE keyboards.

> By convention, for two part split keyboards left "halves" are set as centrals and right halves as peripherals.

Reviewer: ```suggestion
By convention, for a keyboard split into two "halves" the left half is set as the central and the right as a peripheral.
```

When the central has an open slot for a peripheral, it will advertise for connections (which will not be visible to non-ZMK devices).
Then, any peripheral that has not yet bonded to a peripheral will pair to it.
Similar to how [bluetooth profiles](bluetooth.md) are managed between the keyboard and host devices, the bonding information will be stored with the corresponding hardware addresses of the other keyboard part, on both the central and peripheral.

In practice, this means that your split keyboard parts will automatically pair and work the first time they are all on at the same time.
However, if this process somehow went wrong or you used controllers in a different split keyboard configuration before, you will need to explicitly clear the stored bond information so that the parts can pair properly.
For this, please follow [the specified procedure](../troubleshooting/connection-issues.mdx#split-keyboard-halves-unable-to-pair) in the troubleshooting section.

:::warning
If the central keyboard part is either advertising for a pairing or waiting for disconnected peripherals, it will consume more power and drain batteries faster.
:::

## Behaviors with Locality

> Most ZMK [behaviors](../behaviors/index.mdx) are processed exclusively on the central part of the split keyboard since it handles the keymap state and any communication with the host devices.

Reviewer: ```suggestion
Most ZMK [behaviors](../behaviors/index.mdx) are processed exclusively on the central of the split keyboard as it handles the keymap state and any communication with the host devices.
```

> However, certain behaviors have "global" or "source" localities, where they can affect the peripheral parts when invoked.

Reviewer: ```suggestion
However, certain behaviors have "global" or "source" localities, where they can affect the peripherals when invoked.
```

### Global Locality Behaviors

These are behaviors that affect all keyboard parts, such as changing lighting effects:

- [RGB underglow behaviors](../behaviors/underglow.md)
- [Backlight behaviors](../behaviors/backlight.md)
- [Power management behaviors](../behaviors/power.md)
- [Soft off behavior](../behaviors/soft-off.md)

### Source Locality Behaviors

These behaviors only affect the keyboard part that they are invoked from:

- [Reset behaviors](../behaviors/reset.md)

:::warning[Nesting behaviors with locality]
Currently there is [an issue](https://github.com/zmkfirmware/zmk/issues/1494) preventing both global and source locality behaviors from working as expected if they are invoked from another behavior, such as a hold-tap, tap dance or a mod-morph.
For this reason it is recommended that these behaviors are placed directly on a keymap layer.
:::

:::note[Peripheral invocation]

> The peripheral side of the keyboard has to be paired and connected to the central side in order to be able to activate these behaviors, even if it is possible to trigger the behavior using only keys on that side.

Reviewer: ```suggestion
Peripherals must be paired and connected to the central in order to be able to activate these behaviors, even if it is possible to trigger the behavior using only keys on a particular peripheral.
```

This is because the key bindings are processed on the central side which would then instruct the peripheral side to run the behavior's effect.
:::

:::note[Combos]

> [Combos](combos.md) always invoke behaviors with source locality on the central part.

Reviewer: ```suggestion
[Combos](combos.md) always invoke behaviors with source locality on the central.
```
